### PR TITLE
Migrate core to TF2.0

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
part of https://github.com/RasaHQ/research/issues/62 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/policies/embedding_policy.py [hunk 1]
```diff
@@ -638,57 +353,204 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+        dialogue_embed = self._tf_layers["embed.dialogue"](dialogue_transformed)
+
+        return dialogue_embed, mask
+
+    def _embed_label(self, label_in: tf.Tensor) -> tf.Tensor:
+        label = self._tf_layers["ffnn.label"](label_in, self._training)
+        return self._tf_layers["embed.label"](label)
+
+    def batch_loss(self, batch_in: List[tf.Tensor]) -> tf.Tensor:
+        dialogue_in, label_in, _ = batch_in
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]: we should call `self.batch_to_model_data_format` here

File: rasa/core/policies/embedding_policy.py [hunk 2]
```diff
@@ -638,57 +353,204 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+            mask,
+        )
+
+        self.metric_loss.update_state(loss)
+        self.metric_acc.update_state(acc)
+
+        return loss
+
+    def batch_predict(self, batch_in: List[tf.Tensor]) -> Dict[Text, tf.Tensor]:
+        dialogue_in = batch_in[0]
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]: we should call `self.batch_to_model_data_format` here

File: rasa/core/policies/embedding_policy.py [hunk 3]
```diff
@@ -638,57 +353,204 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+        )
+        self._tf_layers["transformer"] = tf_layers.TransformerEncoder(
+            self.config[NUM_TRANSFORMER_LAYERS],
+            self.config[TRANSFORMER_SIZE],
+            self.config[NUM_HEADS],
+            self.config[TRANSFORMER_SIZE] * 4,
+            self.config[MAX_SEQ_LENGTH],
+            self.config[C2],
+            self.config[DROPRATE_DIALOGUE],
+            name="dialogue_encoder",
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]:
```suggestion
            unidirectional=True,
            name="dialogue_encoder",
```

File: rasa/core/policies/embedding_policy.py [hunk 4]
```diff
@@ -638,57 +353,204 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+        return self._tf_layers["embed.label"](label)
+
+    def batch_loss(self, batch_in: List[tf.Tensor]) -> tf.Tensor:
+        dialogue_in, label_in, _ = batch_in
+
+        if self.max_history_tracker_featurizer_used:
+            # add time dimension if max history featurizer is used
+            label_in = label_in[:, tf.newaxis, :]
+
+        all_label, self.all_labels_embed = self._create_all_labels_embed()
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]: there is no need for `self.` in training, it is recalculated every time anyhow
```suggestion
        all_label, all_labels_embed = self._create_all_labels_embed()
```

File: rasa/core/policies/embedding_policy.py [hunk 5]
```diff
@@ -638,57 +353,204 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+
+        dialogue_embed, mask = self._emebed_dialogue(dialogue_in)
+        label_embed = self._embed_label(label_in)
+
+        loss, acc = self._tf_layers["loss.label"](
+            dialogue_embed,
+            label_embed,
+            label_in,
+            self.all_labels_embed,
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]:
```suggestion
            all_labels_embed,
```

File: rasa/core/policies/embedding_policy.py [hunk 6]
```diff
@@ -638,57 +353,204 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+    def _create_all_labels_embed(self) -> Tuple[tf.Tensor, tf.Tensor]:
+        all_label = tf.constant(
+            self._encoded_all_label_ids, dtype=tf.float32, name="all_label"
+        )
+        all_labels_embed = self._embed_label(all_label)
 
-            sim_all = train_utils.load_tensor("similarity_all")
-            pred_confidence = train_utils.load_tensor("pred_confidence")
-            sim = train_utils.load_tensor("similarity")
+        return all_label, all_labels_embed
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]: in nlu, we output embed first, let's use the same style: change here or there

File: rasa/core/policies/keras_policy.py [hunk 7]
```diff
@@ -18,11 +18,13 @@
 from rasa.core.policies.policy import Policy
 from rasa.core.trackers import DialogueStateTracker
 from rasa.utils.common import obtain_verbosity
+from rasa.utils.tensorflow import tf_models
```
Comment (Ghostvv) [inline rasa/core/policies/keras_policy.py]: why do we need it here?

Comment (Ghostvv) [reply rasa/core/policies/keras_policy.py]: our model cannot be used

File: rasa/core/policies/embedding_policy.py [hunk 8]
```diff
@@ -638,57 +353,204 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+            "label",
+            self.config[SIMILARITY_TYPE],
+        )
 
-            a_in = train_utils.load_tensor("user_placeholder")
-            b_in = train_utils.load_tensor("bot_placeholder")
+    def _create_all_labels_embed(self) -> Tuple[tf.Tensor, tf.Tensor]:
+        all_label = tf.constant(
+            self._encoded_all_label_ids, dtype=tf.float32, name="all_label"
+        )
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]: I don't think we need that, numpy array can be used as input to keras layer

File: rasa/core/policies/keras_policy.py (hunk 7 above)
Comment (tabergma) [reply rasa/core/policies/keras_policy.py]: it is not used, will remove it

File: rasa/core/policies/embedding_policy.py [hunk 9]
```diff
@@ -435,199 +232,116 @@ def train(
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]: let's create `label_data` object as we do in nlu

File: rasa/core/policies/embedding_policy.py [hunk 10]
```diff
@@ -638,57 +352,213 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+            self.config[EMBED_DIM],
+            self.config[C2],
+            "label",
+            self.config[SIMILARITY_TYPE],
+        )
 
-        if not os.path.exists(checkpoint + ".meta"):
-            return cls(featurizer=featurizer)
+    def _create_all_labels_embed(self) -> Tuple[np.ndarray, tf.Tensor]:
+        all_label = self._encoded_all_label_ids.astype(np.float32)
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]:
```suggestion
        all_labels = self._encoded_all_label_ids.astype(np.float32)
```

File: rasa/core/policies/embedding_policy.py [hunk 11]
```diff
@@ -638,57 +352,213 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+        all_labels_embed = self._embed_label(all_label)
 
-        meta_file = os.path.join(path, "embedding_policy.json")
-        meta = json.loads(rasa.utils.io.read_file(meta_file))
+        return all_label, all_labels_embed
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]:
```suggestion
        return all_labels, all_labels_embed
```

File: rasa/utils/tensorflow/tf_models.py [hunk 12]
```diff
@@ -291,8 +280,7 @@ def _should_evaluate(
 
     @staticmethod
     def batch_to_model_data_format(
-        batch: Union[Tuple[np.ndarray], Tuple[tf.Tensor]],
-        data_signature: Dict[Text, List[FeatureSignature]],
+        batch: List[tf.Tensor], data_signature: Dict[Text, List[FeatureSignature]]
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/tf_models.py]: it can be `np.ndarray` (e.g. when we calculate `tf_label_data`) also all the batch functions will work with `np.ndarray` as well, let's keep it in type

File: rasa/core/policies/embedding_policy.py [hunk 13]
```diff
@@ -435,199 +232,117 @@ def train(
...
+            model_data,
+            self.config[EPOCHS],
+            self.config[BATCH_SIZES],
+            self.config[EVAL_NUM_EXAMPLES],
+            self.config[EVAL_NUM_EPOCHS],
+            batch_strategy=self.config[BATCH_STRATEGY],
+            random_seed=self.config[RANDOM_SEED],
+        )
 
     def continue_training(
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/core/policies/embedding_policy.py]: I guess it will be removed in the 2.0 release. The function is still called, just a warning is printed.
See https://github.com/RasaHQ/rasa/blob/master/rasa/core/agent.py#L596.

**Comment by tabergma** [comment]
Updated the code so far. A bunch of tests are still failing as they are using old methods, variables. Will fix them before we merge.

**Comment by Ghostvv** [comment]
let's also figure out core prediction accuracy before we merge

**Comment by Ghostvv** [comment]
core accuracy seems to be fine

File: rasa/core/policies/embedding_policy.py [hunk 14]
```diff
@@ -638,57 +358,219 @@ def load(cls, path: Text) -> "EmbeddingPolicy":
...
+            label_data = pickle.load(f)
+
+        with open(os.path.join(path, file_name + ".meta.pkl"), "rb") as f:
+            meta = pickle.load(f)
+
+        if meta[SIMILARITY_TYPE] == "auto":
+            if meta[LOSS_TYPE] == "softmax":
+                meta[SIMILARITY_TYPE] = "inner"
+            elif meta[LOSS_TYPE] == "margin":
+                meta[SIMILARITY_TYPE] = "cosine"
```
Comment (Ghostvv) [inline rasa/core/policies/embedding_policy.py]: we do it so many times, shall we refactor it into the global function?

**Comment by tabergma** [comment]
tests might already pass, let's wait for the build, if there are still a couple of tests failing, then we can do it in a separate PR
